proposals/0167-loader-v4.md

```diff
@@ -0,0 +1,477 @@
+- Behavior:
+  - Charge 32 + chunk_length_in_bytes / cpi_bytes_per_unit CUs
+  - Check there are at least two instruction accounts,
+  otherwise throw `NotEnoughAccountKeys`
+  - Verify the program account header
+  - Check the status stored in the program account is `NeverBeenDeployed` or
+  `Retracted`,
+  otherwise throw `InvalidArgument`
+  - Check that the end offset (sum of offset and length of the chunk) does
+  not exceed the maximum (program account length minus the header size),
+  otherwise throw `AccountDataTooSmall`
+  - Copy the chunk into the program account at the offset shifted by the
+  header size
+
+#### Copy
+
+- Instruction accounts:
+  - `[writable]` The program account to copy to.
+  - `[signer]` The authority of the program.
+  - `[]` The account to copy from.
+- Instruction data:
+  - Enum variant `5u32`
+  - `u32` Byte offset at which to write
+  - `u32` Byte offset at which to read
+  - `u32` Length of the chunk to copy in bytes
+- Behavior:
+  - Charge 32 + chunk_length_in_bytes / cpi_bytes_per_unit CUs
+  - Check there are at least three instruction accounts,
+  otherwise throw `NotEnoughAccountKeys`
+  - Check that program account and source account do not alias,
+  otherwise throw `AccountBorrowFailed`
+  - Verify the program account header
+  - Check the status stored in the program account is `NeverBeenDeployed` or
+  `Retracted`,
+  otherwise throw `InvalidArgument`
+  - Check that the source accounts owner and look-up the source header size:
+    - loader-v2: 0 bytes
+    - loader-v3 buffer: 37 bytes
+    - loader-v3 programdata: 45 bytes
+    - loader-v4: 48 bytes
+    - if none of the above matches throw `InvalidArgument`
+  - Check that the source end offset (sum of source offset and length) does
+  not exceed the maximum (source account length minus the source header size),
+  otherwise throw `AccountDataTooSmall`
+  - Check that the destination end offset (sum of destination offset and
+  length) does not exceed the maximum (program account length minus the loader-v4
+  header size), otherwise throw `AccountDataTooSmall`
+  - Copy the chunk between the program accounts at the offsets, each shifted by
+  the header size of their loader (account owner) respectively
+
+#### Deploy
+
+- Instruction accounts:
+  - `[writable]` The program account to deploy.
+  - `[signer]` The authority of the program.
+- Instruction data:
+  - Enum variant `6u32`
+  - `u32` The executable length in bytes.
+- Behavior:
+  - Charge 32 CUs
+  - Check there are at least two instruction accounts,
+    otherwise throw `NotEnoughAccountKeys`
+  - Verify the program account header
+  - Check that the status stored in the program account is `NeverBeenDeployed`
+  or `Retracted`
+    otherwise throw `InvalidArgument`
+  - Check that the executable length is not longer than the program account,
+    otherwise throw `InvalidArgument`
+  - Charge program_length_in_bytes / cpi_bytes_per_unit CUs
```

Comment:
We should only charge for the executable length rather than the full account data length

```suggestion
  - Charge executable length / cpi_bytes_per_unit CUs
```